ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Restore.hs

```diff
@@ -0,0 +1,175 @@
+import Ariadne.Wallet.Cardano.Kernel.Bip39 (mnemonicToSeedNoPassword)
+import Ariadne.Wallet.Cardano.Kernel.Bip44
+  (Bip44DerivationPath(..), bip44PathToAddressId, decodeBip44DerivationPath)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+  (AssuranceLevel(..), HdAddressId, WalletName(..), eskToHdRootId,
+  hdAddressIdParent)
+import Ariadne.Wallet.Cardano.Kernel.PrefilterTx
+  (PrefilteredUtxo, UtxoByAccount)
+import Ariadne.Wallet.Cardano.Kernel.Wallets
+  (HasNonemptyPassphrase(..), mkHasPP)
+import Ariadne.Wallet.Face (Mnemonic(..))
+
+newtype WrongMnemonic = WrongMnemonic Text
+  deriving (Eq, Show)
+
+instance Exception WrongMnemonic where
+  displayException (WrongMnemonic txt) =
+    "Wrong mnemonic: " <> show txt
+
+data SecretsDecodingError = SecretsDecodingError FilePath Text
+  deriving (Eq, Show)
+
+instance Exception SecretsDecodingError where
+  displayException (SecretsDecodingError path txt) =
+    "Failed to decode " <> path <> ": " <> show txt
+
+data RestoreFrom
+  = RestoreFromMnemonic !(Either Mnemonic BackupPhrase)
+                        !PassPhrase
+  | RestoreFromKeyFile  !FilePath
+
+
+data WalletToRestore = WalletToRestore
+    !EncryptedSecretKey
+    !HasNonemptyPassphrase
+    !WalletName
+    !AssuranceLevel
+
+restoreWallets
+    :: RestoreFrom
+    -> WalletName
+    -> IO [WalletToRestore]
+restoreWallets rFrom walletName = do
+  case rFrom of
+    RestoreFromMnemonic eMnemonicBPhrase pp -> do
+      walletRestore <- restoreWalletFromMnemonic pp walletName eMnemonicBPhrase
+      pure [walletRestore]
+    RestoreFromKeyFile path ->
+      restoreFromKeyFile walletName path
+
+restoreWalletFromMnemonic
+  :: MonadThrow m
+  => PassPhrase
+  -> WalletName
+  -> Either Mnemonic BackupPhrase
+  -> m WalletToRestore
+restoreWalletFromMnemonic pp walletName eMnemonicBPhrase = do
+  let hasPP = mkHasPP pp
+  esk <- getKeyFromMnemonic pp eMnemonicBPhrase
+  pure $ WalletToRestore esk hasPP walletName assurance
+  where
+    -- TODO(AD-251): allow selecting assurance.
+    assurance = AssuranceLevelNormal
+
+
+restoreFromKeyFile
+    :: WalletName
+    -> FilePath
+    -> IO [WalletToRestore]
+restoreFromKeyFile walletName path = do
+    esks <- readNonAriadneKeys path
+    let templateName i (WalletName n) = WalletName $ n <> " " <> pretty i
+    let hasPP esk = HasNonemptyPassphrase $ isNothing $ Crypto.checkPassMatches Crypto.emptyPassphrase esk
+    pure $
+      fmap
+        (\(i,esk) -> WalletToRestore esk (hasPP esk) (templateName i walletName) assurance)
+        (zip [(0 :: Int)..] esks)
+    where
+      -- TODO(AD-251): allow selecting assurance.
+      assurance = AssuranceLevelNormal
+
+getKeyFromMnemonic
+    :: MonadThrow m
+    => PassPhrase
+    -> Either Mnemonic BackupPhrase
+    -> m EncryptedSecretKey
+getKeyFromMnemonic pp = \case
+    Left (Mnemonic mnemonic) ->
+      let mnemonicWords = words mnemonic
+          seed = mnemonicToSeedNoPassword (unwords $ Unsafe.init mnemonicWords)
```

Comment:
I think wallet kernel shouldn't know about this `Unsafe.init` thing, let's do it outside this module.